### PR TITLE
Change the rate limit to be IP/username based to help university students

### PR DESCRIPTION
**Pull request by OmarIthawi**
Task: [The User account is being locked after 5 times of unsuccessful attempts and the validation message does not indicate when the user email will unlock](https://app.asana.com/0/23241728739082/53483528445693)

### CODE REVIEW

**Comment by ahmedaljazzar** [comment]
@OmarIthawi, we are not using Captcha here, no?

**Comment by ahmedaljazzar** [comment]
Ok, It works fine, but why do not we specify how long the learner should wait before trying again? I became confused how much I should wait until I read the code. I do not assume she'll read the code to know when to retry again. :P

So instead of: 

> محاولات فاشلة كثيرة لتسجيل الدخول. يُرجى إعادة المحاولة لاحقًا.
we put:

> محاولات فاشلة كثيرة لتسجيل الدخول. يُرجى إعادة المحاولة بعد ١٥ دقيقة.

**Comment by ahmedaljazzar** [comment]
Ok, It works fine, but why do not we specify how long the learner should wait before trying again? I became confused how much I should wait until I read the code. I do not assume she'll read the code to know when to retry again. :P

So instead of: 

> محاولات فاشلة كثيرة لتسجيل الدخول. يُرجى إعادة المحاولة لاحقًا.

we put:

> محاولات فاشلة كثيرة لتسجيل الدخول. يُرجى إعادة المحاولة بعد ١٥ دقيقة.

File: common/djangoapps/util/bad_request_rate_limiter.py [hunk 1]
```diff
@@ -2,7 +2,31 @@
 A utility class which wraps the RateLimitMixin 3rd party class to do bad request counting
 which can be used for rate limiting
 """
-from ratelimitbackend.backends import RateLimitMixin
+from ratelimitbackend.backends import RateLimitMixin, RateLimitModelBackend
+
+
+class UsernameIPRateLimitModelBackend(RateLimitModelBackend):
+    """
+    Adds username prefix to the original IP-based key.
```
Comment (ahmedaljazzar) [inline common/djangoapps/util/bad_request_rate_limiter.py]: Liked the idea 👍

**Comment by OmarIthawi** [comment]
@ahmedaljazzar I'm still not sure why it doesn't work on beta right (I mean the original edX configs).

Don't merge until I confirm the issue.

**Comment by OmarIthawi** [comment]
> @OmarIthawi, we are not using Captcha here, no?
No we are not. Maybe we add this later on. I don't like to optimize prematurely.

**Comment by ahmedaljazzar** [comment]
When I'm trying to re-login during the 15 mins, It renewed, does that makes sense?

**Comment by OmarIthawi** [comment]
Yes, the original package logic is, that every failed attempts expires after 15 minute. So when it logged you in, you probably had about 10 attempts since the first 5 have already expired. 

The code dictates this.

**Comment by ahmedaljazzar** [comment]
Then here's what I think: We need to inform the learner of how much attempts left, and how long he must wait, I do not prefer keeping him in the dark side.

**Comment by OmarIthawi** [comment]
@ahmedaljazzar Ready for another trial.

**Comment by ahmedaljazzar** [comment]
As discussed earlier:

- [ ] Try to move the Edraak rate limit by ip inside Util.
- [ ] We do not need this [line](https://github.com/Edraak/edx-platform/pull/172/files#diff-e2ffffdf6d67c9e0dee5bc10a67eb8bfR692).

**Comment by ahmedaljazzar** [comment]
As discussed earlier:

- [ ] Try to move the Edraak rate limit by ip inside Util.
- [ ] We do not need this [change](https://github.com/Edraak/edx-platform/pull/172/files#diff-e2ffffdf6d67c9e0dee5bc10a67eb8bfR692).

**Comment by OmarIthawi** [comment]
>  Try to move the Edraak rate limit by ip inside Util.

I can't, it's kinda set in stone.

> We do not need this [change](https://github.com/Edraak/edx-platform/pull/172/files#diff-e2ffffdf6d67c9e0dee5bc10a67eb8bfR692).

Actually, we do. It's for password-reset and other features. I need to loosen up the limit a bit as well!

**Comment by OmarIthawi** [comment]
@ahmedaljazzar the comments are done.

One thing though,

> Ok, It works fine, but why do not we specify how long the learner should wait before trying again? I became confused how much I should wait until I read the code. I do not assume she'll read the code to know when to retry again. :P
> 
> So instead of: 
> 
> > محاولات فاشلة كثيرة لتسجيل الدخول. يُرجى إعادة المحاولة لاحقًا.
> 
> we put:
> 
> > محاولات فاشلة كثيرة لتسجيل الدخول. يُرجى إعادة المحاولة بعد ١٥ دقيقة.

I'd rather to fix this from Transifex, I don't feel like changing the text. What do you think?

This means that the Arabic and the English text will be inconsistent.

**Comment by OmarIthawi** [comment]
@ahmedaljazzar done one.

**Comment by ahmedaljazzar** [comment]
@OmarIthawi LGTM 👍
Merging...

**Pull request merged**